{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Exercise 3\n",
    "\n",
    "### - Chosen datasets: ```Blood stem cell (GSE18067)```, ```Erythroid (GSE18067)```"
   ],
   "id": "50ed7a86f04223d2"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "####  Import libraries",
   "id": "1a991780762baf74"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T18:06:48.588380Z",
     "start_time": "2024-08-21T18:06:48.584345Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import GEOparse\n",
    "import pandas as pd"
   ],
   "id": "76139c70d5445188",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Load the dataset",
   "id": "18fc49ef672d0505"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T17:16:08.642173Z",
     "start_time": "2024-08-21T17:15:23.094571Z"
    }
   },
   "cell_type": "code",
   "source": "gse = GEOparse.get_GEO(geo=\"GSE18067\", destdir=\"data/\")",
   "id": "9b5cc44ccfb63839",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21-Aug-2024 20:15:23 DEBUG utils - Directory data/ already exists. Skipping.\n",
      "21-Aug-2024 20:15:23 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE18nnn/GSE18067/soft/GSE18067_family.soft.gz to data/GSE18067_family.soft.gz\n",
      "100%|██████████| 71.5M/71.5M [00:29<00:00, 2.55MB/s]  \n",
      "21-Aug-2024 20:15:53 DEBUG downloader - Size validation passed\n",
      "21-Aug-2024 20:15:53 DEBUG downloader - Moving /var/folders/6b/mjfxydrs1hb4_cf4_vhhd8lc0000gn/T/tmpfl1a4oq2 to /Users/shir/Documents/PycharmProjects/personal-medicine-tau/exercise3/data/GSE18067_family.soft.gz\n",
      "21-Aug-2024 20:15:53 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE18nnn/GSE18067/soft/GSE18067_family.soft.gz\n",
      "21-Aug-2024 20:15:53 INFO GEOparse - Parsing data/GSE18067_family.soft.gz: \n",
      "21-Aug-2024 20:15:53 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "21-Aug-2024 20:15:53 DEBUG GEOparse - SERIES: GSE18067\n",
      "21-Aug-2024 20:15:53 DEBUG GEOparse - PLATFORM: GPL6238\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451699\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451700\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451701\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451702\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451703\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451704\n",
      "21-Aug-2024 20:15:54 DEBUG GEOparse - SAMPLE: GSM451705\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451706\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451707\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451708\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451709\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451710\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451711\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451712\n",
      "21-Aug-2024 20:15:55 DEBUG GEOparse - SAMPLE: GSM451713\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451714\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451715\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451716\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451717\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451718\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451719\n",
      "21-Aug-2024 20:15:56 DEBUG GEOparse - SAMPLE: GSM451720\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451721\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451722\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451723\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451724\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451725\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451726\n",
      "21-Aug-2024 20:15:57 DEBUG GEOparse - SAMPLE: GSM451727\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451728\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451729\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451730\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451731\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451732\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451733\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451734\n",
      "21-Aug-2024 20:15:58 DEBUG GEOparse - SAMPLE: GSM451735\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451736\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451737\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451738\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451739\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451740\n",
      "21-Aug-2024 20:15:59 DEBUG GEOparse - SAMPLE: GSM451741\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451742\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451743\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451744\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451745\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451746\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451747\n",
      "21-Aug-2024 20:16:00 DEBUG GEOparse - SAMPLE: GSM451748\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451749\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451750\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451751\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451752\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451753\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451754\n",
      "21-Aug-2024 20:16:01 DEBUG GEOparse - SAMPLE: GSM451755\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451756\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451757\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451758\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451759\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451760\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451761\n",
      "21-Aug-2024 20:16:02 DEBUG GEOparse - SAMPLE: GSM451762\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451763\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451764\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451765\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451766\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451767\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451768\n",
      "21-Aug-2024 20:16:03 DEBUG GEOparse - SAMPLE: GSM451769\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451770\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451771\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451772\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451773\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451774\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451775\n",
      "21-Aug-2024 20:16:04 DEBUG GEOparse - SAMPLE: GSM451776\n",
      "21-Aug-2024 20:16:05 DEBUG GEOparse - SAMPLE: GSM451777\n",
      "21-Aug-2024 20:16:05 DEBUG GEOparse - SAMPLE: GSM451778\n",
      "21-Aug-2024 20:16:05 DEBUG GEOparse - SAMPLE: GSM451779\n",
      "21-Aug-2024 20:16:05 DEBUG GEOparse - SAMPLE: GSM451780\n",
      "21-Aug-2024 20:16:05 DEBUG GEOparse - SAMPLE: GSM451781\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451782\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451783\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451784\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451785\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451786\n",
      "21-Aug-2024 20:16:06 DEBUG GEOparse - SAMPLE: GSM451787\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451788\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451789\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451790\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451791\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451792\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451793\n",
      "21-Aug-2024 20:16:07 DEBUG GEOparse - SAMPLE: GSM451794\n",
      "21-Aug-2024 20:16:08 DEBUG GEOparse - SAMPLE: GSM451795\n",
      "21-Aug-2024 20:16:08 DEBUG GEOparse - SAMPLE: GSM451796\n",
      "21-Aug-2024 20:16:08 DEBUG GEOparse - SAMPLE: GSM451797\n",
      "21-Aug-2024 20:16:08 DEBUG GEOparse - SAMPLE: GSM451798\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Inspect the dataset",
   "id": "b62fc68fc6b1dacc"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T20:41:46.745961Z",
     "start_time": "2024-08-21T20:41:46.706866Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(\"GSM example:\")\n",
    "for gsm_name, gsm in gse.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print (\"Table data:\",)\n",
    "    print (gsm.table.head())\n",
    "    break\n",
    "\n",
    "print()\n",
    "print(\"GPL example:\")\n",
    "for gpl_name, gpl in gse.gpls.items():\n",
    "    print(\"Name: \", gpl_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gpl.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print(\"Table data:\",)\n",
    "    print(gpl.table.head())\n",
    "    break"
   ],
   "id": "9f111b7b7aa7d718",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM example:\n",
      "Name:  GSM451699\n",
      "Metadata:\n",
      " - title : BXD13 Myeloid batch1\n",
      " - geo_accession : GSM451699\n",
      " - status : Public on Sep 12 2009\n",
      " - submission_date : Sep 10 2009\n",
      " - last_update_date : Sep 11 2009\n",
      " - type : RNA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : Myeloid\n",
      " - organism_ch1 : Mus musculus\n",
      " - taxid_ch1 : 10090\n",
      " - characteristics_ch1 : mouse id: 13, bxd strain: 13, cell type: Gr1+, array batch: 1\n",
      " - molecule_ch1 : total RNA\n",
      " - extract_protocol_ch1 : Total RNA was isolated using the RNeasy Mini kit (Qiagen) in accordance with the manufacturer’s protocol.\n",
      " - label_ch1 : biotin\n",
      " - label_protocol_ch1 : Biotinylated cRNA was prepared using the Illumina TotalPrep RNA Amplification Kit (Ambion) according to the manufacturer’s specifications starting with 100 ng total RNA\n",
      " - hyb_protocol : Standard Illumina hybridization protocol\n",
      " - scan_protocol : Standard Illumina scanning protocol\n",
      " - description : none\n",
      " - data_processing : Quantile normalization was applied on log2 transformed data.\n",
      " - platform_id : GPL6238\n",
      " - contact_name : Bruno,,Tesson\n",
      " - contact_laboratory : GBiC\n",
      " - contact_department : Groningen Bioinformatics Center\n",
      " - contact_institute : University of Groningen\n",
      " - contact_address : Kerklaan 30\n",
      " - contact_city : Haren\n",
      " - contact_zip/postal_code : 9750NN\n",
      " - contact_country : Netherlands\n",
      " - supplementary_file : NONE\n",
      " - series_id : GSE18067\n",
      " - data_row_count : 46628\n",
      "Table data:\n",
      "              ID_REF     VALUE  Detection Pval\n",
      "0  10181072_239_rc-S  6.549379        0.592639\n",
      "1  10181072_290_rc-S  6.454085        0.774173\n",
      "2     10181072_290-S  6.800331        0.165315\n",
      "3  10181072_311_rc-S  7.040595        0.051778\n",
      "4     10181072_311-S  6.556179        0.577043\n",
      "\n",
      "GPL example:\n",
      "Name:  GPL6238\n",
      "Metadata:\n",
      " - title : mouse-6 v1.1 (Illumina)\n",
      " - geo_accession : GPL6238\n",
      " - status : Public on Feb 11 2008\n",
      " - submission_date : Dec 04 2007\n",
      " - last_update_date : Feb 10 2009\n",
      " - technology : oligonucleotide beads\n",
      " - distribution : commercial\n",
      " - organism : Mus musculus\n",
      " - taxid : 10090\n",
      " - manufacturer : Illumina Inc.\n",
      " - manufacture_protocol : see manufacturer's website\n",
      " - description : This platform represents the commercially-available Illumina mouse-6 v1.1 expression beadchip, but is not intended for general use., Please contact the GEO staff if you need help submitting Illumina data.\n",
      " - contact_name : GEO,,admin\n",
      " - contact_email : geo@ncbi.nlm.nih.gov\n",
      " - contact_institute : NCBI/NLM/NIH\n",
      " - contact_address : 9000 Rockville Pike\n",
      " - contact_city : Bethesda\n",
      " - contact_state : MD\n",
      " - contact_zip/postal_code : 20892\n",
      " - contact_country : USA\n",
      " - relation : Alternative to: GPL6105 (official)\n",
      " - data_row_count : 46628\n",
      "Table data:\n",
      "                   ID        Search_key      ProbeId          Gid  \\\n",
      "0       GI_38090455-S       GI_38090455  105290026.0  GI_38090455   \n",
      "1  scl0011717.2_283-S  scl0011717.2_283     610369.0   GI_6753047   \n",
      "2  scl52892.1.1_103-S  scl52892.1.1_103    1450014.0  GI_31342363   \n",
      "3    scl0012946.2_3-S    scl0012946.2_3     380019.0   GI_7304978   \n",
      "4    scl18939.11_88-S    scl18939.11_88  100430441.0   GI_6681286   \n",
      "\n",
      "         Transcript       GB_ACC         Symbol Type   Start  \\\n",
      "0       GI_38090455  XM_356462.1      LOC382362    S   760.0   \n",
      "1  scl0011717.2_283  NM_009667.1          Ampd3    S  1979.0   \n",
      "2  scl52892.1.1_103  NM_176907.2  D830016O14Rik    S  1919.0   \n",
      "3    scl0012946.2_3  NM_013499.1           Crry    S   400.0   \n",
      "4    scl18939.11_88    NM_007914            Ehf    S    17.0   \n",
      "\n",
      "                                            SEQUENCE  \\\n",
      "0  GATGGTGACCGTCTTATCACAGAACCTCTGCCGGAACTACAGCTGG...   \n",
      "1  ACCGATGACCCCATGCAGTTTCACTACACAAAGAAGCGCTCATGGA...   \n",
      "2  GGTGTCTTTAACAAATCCCTCCAAAGTGAGCTTAATTTGACAAACA...   \n",
      "3  CGGAACCCAGCGCCATTTCTGGGTGGGACTGCTTTCTACACCATTT...   \n",
      "4  CCACCCTACAGACTAAATGTTGTAACAGAACATGGACAGCTTTGGT...   \n",
      "\n",
      "                                          Definition  \\\n",
      "0  Mus musculus similar to heparan sulfate (gluco...   \n",
      "1        Mus musculus AMP deaminase 3 (Ampd3), mRNA.   \n",
      "2  Mus musculus RIKEN cDNA D830016O14 gene (D8300...   \n",
      "3  Mus musculus complement receptor related prote...   \n",
      "4    Mus musculus ets homologous factor (Ehf), mRNA.   \n",
      "\n",
      "                                            Ontology        Synonym SPOT_ID  \n",
      "0                                                NaN            NaN     NaN  \n",
      "1  go_function: hydrolase activity, acting on car...            NaN     NaN  \n",
      "2                                                NaN            NaN     NaN  \n",
      "3  go_function: receptor activity [goid 0004872] ...       Mcp;mCRY     NaN  \n",
      "4  go_component: nucleus [goid 0005634] [evidence...  9030625L19Rik     NaN  \n"
     ]
    }
   ],
   "execution_count": 66
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Extract expression data and annotations",
   "id": "396d47a94fa1db93"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T22:09:54.349627Z",
     "start_time": "2024-08-21T22:09:53.405963Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# extract the samples and platform data\n",
    "samples = gse.gsms\n",
    "platform = gse.gpls['GPL6238']\n",
    "\n",
    "# extract the expression data and convert into a DataFrame\n",
    "expression_data = {}\n",
    "for sample_name, sample in samples.items():\n",
    "    title = sample.metadata['title'][0]\n",
    "    strain = title.split(' ')[0]\n",
    "    cell_type = title.split(' ')[1]\n",
    "    technical_info_list = title.split(' ')[2:]\n",
    "    technical_info = '_'.join(technical_info_list)\n",
    "    \n",
    "    # add the cell type to the expression data if it doesn't exist\n",
    "    if cell_type not in expression_data.keys():\n",
    "        expression_data[cell_type] = {}\n",
    "    \n",
    "    # add the strain to the cell type data\n",
    "    expression_data[cell_type][f'{strain}_{technical_info}'] = sample.table.set_index('ID_REF')['VALUE']\n",
    "        \n",
    "# convert to DF\n",
    "for cell_type, cell_data in expression_data.items():\n",
    "    expression_data[cell_type] = pd.DataFrame(cell_data)"
   ],
   "id": "84aadacbda2dd25d",
   "outputs": [],
   "execution_count": 116
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "##### Merge with the platform annotations to get gene symbols or Entrez IDs",
   "id": "9f5c52d67d94f105"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T22:10:00.252296Z",
     "start_time": "2024-08-21T22:10:00.096137Z"
    }
   },
   "cell_type": "code",
   "source": [
    "annotation_df = platform.table[['ID', 'Symbol']].copy()\n",
    "\n",
    "# Merge the expression data with the annotation data\n",
    "for cell_type, expression_df in expression_data.items():\n",
    "    # name columns by strain\n",
    "    expression_df.columns = [col.split('_')[0] for col in expression_df.columns]\n",
    "    \n",
    "    # merge\n",
    "    merged_df = expression_df.merge(annotation_df, left_index=True, right_on='ID')\n",
    "    \n",
    "    # Set the gene symbol as the index and drop the ID column\n",
    "    merged_df = merged_df.set_index('Symbol').drop('ID', axis=1)\n",
    "    \n",
    "    # Save the merged data\n",
    "    expression_data[cell_type] = merged_df"
   ],
   "id": "2c1457de7352f36",
   "outputs": [],
   "execution_count": 117
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "\n",
    "\n",
    "### Data Preprocessing"
   ],
   "id": "e056803b005c95c0"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "#### Define the cell types of the analysis\n",
    "The variable ```cell_types``` should be a list containing two of the following: ```['Progenitor', 'Stem', 'Erythroid', 'Myeloid']```"
   ],
   "id": "b3278f2a74699375"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T22:10:04.796286Z",
     "start_time": "2024-08-21T22:10:04.789638Z"
    }
   },
   "cell_type": "code",
   "source": [
    "cell_types = [\"Stem\", \"Erythroid\"]\n",
    "\n",
    "# get expression data for the desired cell types\n",
    "preprocessed_data = {}\n",
    "for cell_type in cell_types:\n",
    "    preprocessed_data[cell_type] = expression_data[cell_type]"
   ],
   "id": "b75f62701cf0916a",
   "outputs": [],
   "execution_count": 118
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Preprocess",
   "id": "e5dfb5cbda585fe1"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T22:10:14.994772Z",
     "start_time": "2024-08-21T22:10:08.823803Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# set preprocessing params\n",
    "maximal_val_threshold = 6.75\n",
    "variance_threshold = 0.006\n",
    "\n",
    "# load genotyping data\n",
    "genotypes = pd.read_excel('data/genotypes.xls', skiprows=1, index_col=0, header=0)\n",
    "\n",
    "# preprocess\n",
    "for cell_type, expression_data in preprocessed_data.items():\n",
    "    # remove rows lacking gene identifier\n",
    "    expression_data = expression_data[expression_data.index.notnull()]\n",
    "    \n",
    "    # remove rows with low maximal value\n",
    "    filtered_data = expression_data[expression_data.max(axis=1) > maximal_val_threshold]\n",
    "\n",
    "    # remove rows with low variance\n",
    "    filtered_data = filtered_data[filtered_data.var(axis=1) > variance_threshold]\n",
    "    print(filtered_data)\n",
    "\n",
    "    # for multiple probes of the same gene, keep the one with the highest variance\n",
    "    filtered_data = filtered_data.groupby(filtered_data.index).apply(lambda x: x.loc[x.var(axis=1).idxmax()])\n",
    "\n",
    "    # use only genes that are present in genotypes.xls\n",
    "    #TODO: make sure that I didn't miss anything\n",
    "    #common_strains = filtered_df.columns.intersection(genotype_data.columns)\n",
    "    #filtered_df = filtered_df[common_strains]\n",
    "    #genotype_data = genotype_data[common_strains]\n",
    "\n",
    "    # average gene expression data across individuals of the same strain\n",
    "    #averaged_df = filtered_df.groupby(filtered_df.columns, axis=1).mean()\n",
    "    \n",
    "    # save the processed dataset\n",
    "    #averaged_df.to_csv(f'output/{cell_type}_processed_expression_data.csv')\n",
    "    \n",
    "    print(filtered_data)"
   ],
   "id": "6d27d208c6c3d033",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    BXD6     BXD28a      BXD19      BXD15      BXD40  \\\n",
      "Symbol                                                                 \n",
      "Pigt           14.557293  14.324239  14.500165  14.834139  13.975895   \n",
      "St6galnac3      6.825827   6.949788   7.037665   7.001356   6.922815   \n",
      "Rangnrf        11.559403  10.742308  11.120819  10.725605  10.840271   \n",
      "L1cam           6.677506   6.531395   6.829248   6.719764   6.579074   \n",
      "Olfr159         6.786502   6.639528   6.755817   6.646977   6.677762   \n",
      "...                  ...        ...        ...        ...        ...   \n",
      "4930563H03Rik   6.656988   6.718578   6.604799   6.612627   6.711276   \n",
      "2310051F07Rik  10.790007  10.489667  10.364562  10.104436   9.700218   \n",
      "1110029L17Rik   9.037357   9.066524   8.995373   8.874153   9.089833   \n",
      "1110059G02Rik   8.008045   7.925591   7.582184   7.551539   7.788641   \n",
      "LOC333685       6.780722   6.890108   6.935633   6.878462   6.863248   \n",
      "\n",
      "                   BXD12      BXD31       BXD9      BXD34     BXD28b  ...  \\\n",
      "Symbol                                                                ...   \n",
      "Pigt           14.472226  14.139170  14.783999  13.813608  14.332498  ...   \n",
      "St6galnac3      7.042718   6.873210   6.937862   7.011879   6.993547  ...   \n",
      "Rangnrf        10.699779  10.872730  10.832521  11.977666  11.258417  ...   \n",
      "L1cam           6.660001   6.671821   6.592675   6.517446   6.677799  ...   \n",
      "Olfr159         6.634708   6.557726   6.595044   6.574110   6.692688  ...   \n",
      "...                  ...        ...        ...        ...        ...  ...   \n",
      "4930563H03Rik   6.820784   6.799266   6.764988   6.820784   6.746665  ...   \n",
      "2310051F07Rik  10.323547  10.203940  10.661355  10.293074  10.405218  ...   \n",
      "1110029L17Rik   8.865190   9.212193   9.305834   9.609414   9.021002  ...   \n",
      "1110059G02Rik   7.749219   8.218362   7.666823   7.907893   7.922678  ...   \n",
      "LOC333685       6.966422   6.877841   6.833295   6.864550   6.822174  ...   \n",
      "\n",
      "                  BXD33b      BXD16      BXD18       BXD2      BXD27  \\\n",
      "Symbol                                                                 \n",
      "Pigt           13.926018  14.349152  14.490862  14.472226  14.415247   \n",
      "St6galnac3      6.853104   8.275517   6.878496   6.872244   7.042332   \n",
      "Rangnrf        10.814006  11.581121  10.706610  11.194534  11.062581   \n",
      "L1cam           6.662716   6.794406   6.720661   6.744707   6.552808   \n",
      "Olfr159         6.677839   7.157628   6.612397   6.675205   6.710507   \n",
      "...                  ...        ...        ...        ...        ...   \n",
      "4930563H03Rik   6.693321   6.420971   6.763720   6.755271   6.654735   \n",
      "2310051F07Rik  10.226224  10.917194  10.222152  10.423419  10.715348   \n",
      "1110029L17Rik   9.408644   8.004286   9.135946   8.933257   9.085948   \n",
      "1110059G02Rik   7.740952   7.708820   7.891572   7.711256   8.263904   \n",
      "LOC333685       6.821267   8.282160   6.793561   6.794038   6.820848   \n",
      "\n",
      "                  BXD33b     BXD21B      BXD14       BXD8      BXD29  \n",
      "Symbol                                                                \n",
      "Pigt           13.453633  12.921902  13.003162  13.622500  12.976506  \n",
      "St6galnac3      7.010506   7.178347   6.940478   7.308123   7.311794  \n",
      "Rangnrf        10.647173  10.985071  10.793278  11.710422  11.549042  \n",
      "L1cam           6.770765   6.786070   6.713485   6.782594   6.576214  \n",
      "Olfr159         6.677894   6.688477   6.598918   6.795848   6.524329  \n",
      "...                  ...        ...        ...        ...        ...  \n",
      "4930563H03Rik   6.776569   6.706933   6.686737   6.811972   6.938753  \n",
      "2310051F07Rik  10.028790  10.008683   9.807446  10.478967  10.543521  \n",
      "1110029L17Rik   9.640516   9.459737   9.408210   9.301047   9.703850  \n",
      "1110059G02Rik   7.866076   7.817624   7.920182   8.171814   8.462044  \n",
      "LOC333685       6.950260   6.979174   6.916670   7.314041   7.007740  \n",
      "\n",
      "[29907 rows x 26 columns]\n"
     ]
    },
    {
     "ename": "InvalidIndexError",
     "evalue": "Reindexing only valid with uniquely valued Index objects",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mInvalidIndexError\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[119], line 21\u001B[0m\n\u001B[1;32m     18\u001B[0m \u001B[38;5;28mprint\u001B[39m(filtered_data)\n\u001B[1;32m     20\u001B[0m \u001B[38;5;66;03m# for multiple probes of the same gene, keep the one with the highest variance\u001B[39;00m\n\u001B[0;32m---> 21\u001B[0m filtered_data \u001B[38;5;241m=\u001B[39m filtered_data\u001B[38;5;241m.\u001B[39mgroupby(filtered_data\u001B[38;5;241m.\u001B[39mindex)\u001B[38;5;241m.\u001B[39mapply(\u001B[38;5;28;01mlambda\u001B[39;00m x: x\u001B[38;5;241m.\u001B[39mloc[x\u001B[38;5;241m.\u001B[39mvar(axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m1\u001B[39m)\u001B[38;5;241m.\u001B[39midxmax()])\n\u001B[1;32m     23\u001B[0m \u001B[38;5;66;03m# use only genes that are present in genotypes.xls\u001B[39;00m\n\u001B[1;32m     24\u001B[0m \u001B[38;5;66;03m#TODO: make sure that I didn't miss anything\u001B[39;00m\n\u001B[1;32m     25\u001B[0m \u001B[38;5;66;03m#common_strains = filtered_df.columns.intersection(genotype_data.columns)\u001B[39;00m\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m     32\u001B[0m \u001B[38;5;66;03m# save the processed dataset\u001B[39;00m\n\u001B[1;32m     33\u001B[0m \u001B[38;5;66;03m#averaged_df.to_csv(f'output/{cell_type}_processed_expression_data.csv')\u001B[39;00m\n\u001B[1;32m     35\u001B[0m \u001B[38;5;28mprint\u001B[39m(filtered_data)\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/groupby/groupby.py:1824\u001B[0m, in \u001B[0;36mGroupBy.apply\u001B[0;34m(self, func, include_groups, *args, **kwargs)\u001B[0m\n\u001B[1;32m   1822\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m option_context(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mmode.chained_assignment\u001B[39m\u001B[38;5;124m\"\u001B[39m, \u001B[38;5;28;01mNone\u001B[39;00m):\n\u001B[1;32m   1823\u001B[0m     \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m-> 1824\u001B[0m         result \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_python_apply_general(f, \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_selected_obj)\n\u001B[1;32m   1825\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m (\n\u001B[1;32m   1826\u001B[0m             \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mobj, Series)\n\u001B[1;32m   1827\u001B[0m             \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_selection \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m   1828\u001B[0m             \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_selected_obj\u001B[38;5;241m.\u001B[39mshape \u001B[38;5;241m!=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_obj_with_exclusions\u001B[38;5;241m.\u001B[39mshape\n\u001B[1;32m   1829\u001B[0m         ):\n\u001B[1;32m   1830\u001B[0m             warnings\u001B[38;5;241m.\u001B[39mwarn(\n\u001B[1;32m   1831\u001B[0m                 message\u001B[38;5;241m=\u001B[39m_apply_groupings_depr\u001B[38;5;241m.\u001B[39mformat(\n\u001B[1;32m   1832\u001B[0m                     \u001B[38;5;28mtype\u001B[39m(\u001B[38;5;28mself\u001B[39m)\u001B[38;5;241m.\u001B[39m\u001B[38;5;18m__name__\u001B[39m, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mapply\u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m   1835\u001B[0m                 stacklevel\u001B[38;5;241m=\u001B[39mfind_stack_level(),\n\u001B[1;32m   1836\u001B[0m             )\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/groupby/groupby.py:1889\u001B[0m, in \u001B[0;36mGroupBy._python_apply_general\u001B[0;34m(self, f, data, not_indexed_same, is_transform, is_agg)\u001B[0m\n\u001B[1;32m   1886\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m not_indexed_same \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m   1887\u001B[0m     not_indexed_same \u001B[38;5;241m=\u001B[39m mutated\n\u001B[0;32m-> 1889\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_wrap_applied_output(\n\u001B[1;32m   1890\u001B[0m     data,\n\u001B[1;32m   1891\u001B[0m     values,\n\u001B[1;32m   1892\u001B[0m     not_indexed_same,\n\u001B[1;32m   1893\u001B[0m     is_transform,\n\u001B[1;32m   1894\u001B[0m )\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/groupby/generic.py:1620\u001B[0m, in \u001B[0;36mDataFrameGroupBy._wrap_applied_output\u001B[0;34m(self, data, values, not_indexed_same, is_transform)\u001B[0m\n\u001B[1;32m   1617\u001B[0m         \u001B[38;5;28;01mreturn\u001B[39;00m result\n\u001B[1;32m   1618\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m   1619\u001B[0m     \u001B[38;5;66;03m# values are Series\u001B[39;00m\n\u001B[0;32m-> 1620\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_wrap_applied_output_series(\n\u001B[1;32m   1621\u001B[0m         values,\n\u001B[1;32m   1622\u001B[0m         not_indexed_same,\n\u001B[1;32m   1623\u001B[0m         first_not_none,\n\u001B[1;32m   1624\u001B[0m         key_index,\n\u001B[1;32m   1625\u001B[0m         is_transform,\n\u001B[1;32m   1626\u001B[0m     )\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/groupby/generic.py:1644\u001B[0m, in \u001B[0;36mDataFrameGroupBy._wrap_applied_output_series\u001B[0;34m(self, values, not_indexed_same, first_not_none, key_index, is_transform)\u001B[0m\n\u001B[1;32m   1640\u001B[0m all_indexed_same \u001B[38;5;241m=\u001B[39m all_indexes_same(x\u001B[38;5;241m.\u001B[39mindex \u001B[38;5;28;01mfor\u001B[39;00m x \u001B[38;5;129;01min\u001B[39;00m values)\n\u001B[1;32m   1642\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m all_indexed_same:\n\u001B[1;32m   1643\u001B[0m     \u001B[38;5;66;03m# GH 8467\u001B[39;00m\n\u001B[0;32m-> 1644\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_concat_objects(\n\u001B[1;32m   1645\u001B[0m         values,\n\u001B[1;32m   1646\u001B[0m         not_indexed_same\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m,\n\u001B[1;32m   1647\u001B[0m         is_transform\u001B[38;5;241m=\u001B[39mis_transform,\n\u001B[1;32m   1648\u001B[0m     )\n\u001B[1;32m   1650\u001B[0m \u001B[38;5;66;03m# Combine values\u001B[39;00m\n\u001B[1;32m   1651\u001B[0m \u001B[38;5;66;03m# vstack+constructor is faster than concat and handles MI-columns\u001B[39;00m\n\u001B[1;32m   1652\u001B[0m stacked_values \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39mvstack([np\u001B[38;5;241m.\u001B[39masarray(v) \u001B[38;5;28;01mfor\u001B[39;00m v \u001B[38;5;129;01min\u001B[39;00m values])\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/groupby/groupby.py:1457\u001B[0m, in \u001B[0;36mGroupBy._concat_objects\u001B[0;34m(self, values, not_indexed_same, is_transform)\u001B[0m\n\u001B[1;32m   1454\u001B[0m     group_levels \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_grouper\u001B[38;5;241m.\u001B[39mlevels\n\u001B[1;32m   1455\u001B[0m     group_names \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_grouper\u001B[38;5;241m.\u001B[39mnames\n\u001B[0;32m-> 1457\u001B[0m     result \u001B[38;5;241m=\u001B[39m concat(\n\u001B[1;32m   1458\u001B[0m         values,\n\u001B[1;32m   1459\u001B[0m         axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39maxis,\n\u001B[1;32m   1460\u001B[0m         keys\u001B[38;5;241m=\u001B[39mgroup_keys,\n\u001B[1;32m   1461\u001B[0m         levels\u001B[38;5;241m=\u001B[39mgroup_levels,\n\u001B[1;32m   1462\u001B[0m         names\u001B[38;5;241m=\u001B[39mgroup_names,\n\u001B[1;32m   1463\u001B[0m         sort\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m,\n\u001B[1;32m   1464\u001B[0m     )\n\u001B[1;32m   1465\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m   1466\u001B[0m     \u001B[38;5;66;03m# GH5610, returns a MI, with the first level being a\u001B[39;00m\n\u001B[1;32m   1467\u001B[0m     \u001B[38;5;66;03m# range index\u001B[39;00m\n\u001B[1;32m   1468\u001B[0m     keys \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mlist\u001B[39m(\u001B[38;5;28mrange\u001B[39m(\u001B[38;5;28mlen\u001B[39m(values)))\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/reshape/concat.py:395\u001B[0m, in \u001B[0;36mconcat\u001B[0;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001B[0m\n\u001B[1;32m    380\u001B[0m     copy \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mFalse\u001B[39;00m\n\u001B[1;32m    382\u001B[0m op \u001B[38;5;241m=\u001B[39m _Concatenator(\n\u001B[1;32m    383\u001B[0m     objs,\n\u001B[1;32m    384\u001B[0m     axis\u001B[38;5;241m=\u001B[39maxis,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    392\u001B[0m     sort\u001B[38;5;241m=\u001B[39msort,\n\u001B[1;32m    393\u001B[0m )\n\u001B[0;32m--> 395\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m op\u001B[38;5;241m.\u001B[39mget_result()\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/reshape/concat.py:680\u001B[0m, in \u001B[0;36m_Concatenator.get_result\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m    678\u001B[0m         obj_labels \u001B[38;5;241m=\u001B[39m obj\u001B[38;5;241m.\u001B[39maxes[\u001B[38;5;241m1\u001B[39m \u001B[38;5;241m-\u001B[39m ax]\n\u001B[1;32m    679\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m new_labels\u001B[38;5;241m.\u001B[39mequals(obj_labels):\n\u001B[0;32m--> 680\u001B[0m             indexers[ax] \u001B[38;5;241m=\u001B[39m obj_labels\u001B[38;5;241m.\u001B[39mget_indexer(new_labels)\n\u001B[1;32m    682\u001B[0m     mgrs_indexers\u001B[38;5;241m.\u001B[39mappend((obj\u001B[38;5;241m.\u001B[39m_mgr, indexers))\n\u001B[1;32m    684\u001B[0m new_data \u001B[38;5;241m=\u001B[39m concatenate_managers(\n\u001B[1;32m    685\u001B[0m     mgrs_indexers, \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mnew_axes, concat_axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mbm_axis, copy\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mcopy\n\u001B[1;32m    686\u001B[0m )\n",
      "File \u001B[0;32m~/anaconda3/envs/Problematic-iSNVs_SARS-CoV-2/lib/python3.11/site-packages/pandas/core/indexes/base.py:3885\u001B[0m, in \u001B[0;36mIndex.get_indexer\u001B[0;34m(self, target, method, limit, tolerance)\u001B[0m\n\u001B[1;32m   3882\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_check_indexing_method(method, limit, tolerance)\n\u001B[1;32m   3884\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_index_as_unique:\n\u001B[0;32m-> 3885\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m InvalidIndexError(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_requires_unique_msg)\n\u001B[1;32m   3887\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mlen\u001B[39m(target) \u001B[38;5;241m==\u001B[39m \u001B[38;5;241m0\u001B[39m:\n\u001B[1;32m   3888\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m np\u001B[38;5;241m.\u001B[39marray([], dtype\u001B[38;5;241m=\u001B[39mnp\u001B[38;5;241m.\u001B[39mintp)\n",
      "\u001B[0;31mInvalidIndexError\u001B[0m: Reindexing only valid with uniquely valued Index objects"
     ]
    }
   ],
   "execution_count": 119
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
